### ISSUE TITLE
ConditionalNodes and FieldCollector can't be extended to include customized query directives

### ISSUE THREAD

**Issue by linqi** [graphql-java/graphql-java#1811]
In my use case, I want to introduce some query directives that function similarly to @skip and @include, for example @featureFlag.

In the current implementation, ConditionalNodes and FieldCollector are all sealed variables in their usages and can't be easily extended. In order to do so, I have to create a code copy of lots of classes. 

Can we make FieldCollector extendable for new ConditionalNodes?

**Comment by bbakerman** [comment]
Can you tell us more about your @FeatureFlag directive?  What behaviour do you want?

```
query {
     someField @FeatureFlag
     someOtherField
}
```

What happens when feature flag is present??

**Comment by linqi** [comment]
Something like the code below

`query {
             someField @FeatureFlag(flagName: my-feature-flag, value: expectedTreatment)
}`

My runtime checks the feature flag's actual value from some db/service and compares it with the given `expectedTreatment` in the query. It uses the comparison result to either include/exclude the field.

**Comment by bsara** [comment]
I need this exact same functionality. Also, [the GraphQL spec allows for custom query level directives, not just schema level ones](https://graphql.org/learn/queries/#directives:~:text=Server%20implementations%20may%20also%20add%20experimental%20features%20by%20defining%20completely%20new%20directives.). It would be great if custom query level directives were supported.

**Comment by bbakerman** [comment]
custom query level directives were supported are supported - what you are really asking for is support in the engine on how a custom directive would execute.

It's VERY tricky to develop and engine that allows new behavior to be introduced.  We have tried in the past and it was less than successful.  

@bsara - can you expand your use case?  Is it the feature flag capability above?  Is it field inclusion / exclusion - eg rebranded @skip / @include ?

**Comment by bsara** [comment]
> can you expand your use case?

I want to be able to create a directive like `isFeatureEnabled` which takes a list of features that the back-end would check to see if they are enabled and omit results if they are not enabled. Yes, we could do this check via a boolean based as a variable, but that creates security concerns. Anyone could easily pass `true` and still get the data if that is the only guard in place. If I'm able to hard code the use of the directive into my query, I can then have an allowlist that checks the hashes of queries sent in an underlying proxy and only allow those that we have written explicitly, thus preventing one from bypassing the feature check while allowing for the flexibility for the client to tell the server which features is expects to be available for the data to be returned.

**Comment by bbakerman** [comment]
Thanks for that - this really helps us.

Spitballing how this might work - we would need to provide a way to register a "callback" into the ConditionalNodes code

(rather than inheritance - we would prefer composition)

So that when we use `ConditionalNodes` we would need to make calls to this plugged in code and decide on whether its ok

One thing to note is - not all `graphql.execution.ConditionalNodes` calls are inside a graphql request per se.

We have code like `graphql.analysis.QueryVisitor / graphql.analysis.QueryTraverser` that will follow skip and include today if you pass in variables - but it might not be proper graphql request 

Same with the new `graphql.normalized.ExecutableNormalizedOperationFactory` - it can run without being a graphql request proper

So we **may** need to reflect this maybe in the invocation - eg this IS a graphql engine call and this is not??

**Comment by bbakerman** [comment]
@bsara -

Right now we have

```
public boolean shouldInclude(Map<String, Object> variables, List<Directive> directives)
```

and we could include what AST element (field / fragment or inline fragment) is being included / excluded say but what else would you want during such a callback?

eg what information needs to feed into your decisions.  I am thinking `graphql.GraphQLContext` say if it was during a graphql request proper say

**Comment by dondonz** [comment]
Closed by pull request #3223, thanks for the suggestion!
